src/Microsoft.AspNetCore.Mvc.Modules/ShellFeatureApplicationPart.cs

```diff
@@ -70,5 +86,30 @@ public IEnumerable<string> GetReferencePaths()
 
 			return _referencePaths;
         }
+
+        /// <inheritdoc />
+        private IEnumerable<Type> GetApplicationTypes()
+        {
+            if (_applicationTypes != null)
+            {
+                return _applicationTypes;
+            }
+
+            lock (_synLock)
+            {
+                if (_applicationTypes != null)
+                {
+                    return _applicationTypes;
+                }
+                var hostingEnvironment = _httpContextAccessor.HttpContext.RequestServices.GetRequiredService<IHostingEnvironment>();
+
+                _applicationTypes = DefaultAssemblyPartDiscoveryProvider
```

Comment:
Find application parts, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L77), so only those related to assemblies which have a reference on a primary mvc assembly, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs#L18). Then we use `ExportedTypes` but maybe better to use `DefinedTypes`, [see here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs#L45). So, here we get types as if `AddMvcCore()` was called at the app level.

Then, after we exclude types exposed by features which are not enabled.

Not sure, just a suggestion, but the goal would be to be sure we don't miss any types.

If features Types don't take into account all the dependency graph, maybe better to 1st start from a more complete list, then exclude what is sufficient to inhibit features which are not enabled.